### PR TITLE
[WIP] [Serve] Refactor ServeHandle

### PR DESCRIPTION
**Pull request by simon-mo**
## Current State:
- [x] Core logic code complete
- [ ] User API code complete
- [ ] Pass existing tests
- [x] Add new test cases (for long puller)
- [ ] Address known issues:
   - It should work in detach & non-detached mode
   - Address scale down race condition*
- [ ] Performance QA

### Scale down race condition
Due to long pulling we have a new race condition when killing a worker:
1. Controller decide to kill a replica, called `ray.kill(actor_handle)`.
2. ServeHandles are long pulling the notification that worker replica was removed.
3. Before they received this message, they can still send some queries to the replica.
4. These queries will arrive after the worker is killed. Therefore immediately errors.

Proposed solution: 
- Instead of asking ray.kill, try to see if we can terminate actor gracefully _inside_ the replica.
- Basically, the replica will wait for all queries to finish and then wait another 10s, then kill itself.

### CODE REVIEW

File: python/ray/serve/tests/test_dev.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import asyncio
```
Comment (simon-mo) [inline python/ray/serve/tests/test_dev.py:1]: ignore this file, it's used for in development testing.

**Comment by edoakes** [comment]
@simon-mo re: the shutdown issue, the proposed solution for "graceful shutdown" of workers sounds good to me though we should make sure it isn't reliant on timing if possible.

We should also be able to just retry the queries when they fail, right? Though I guess an issue here is that we can't currently distinguish between a worker crashing (will be restarted) and when it's intentionally `ray.kill`'d, right?

**Comment by edoakes** [comment]
@simon-mo at first glance, it looks like there are basically two unrelated changes: changing the router to long-poll for updates and also refactoring the handle to encompass the router. In its current state this is only doing the former, right? I would prefer if we do these in two separate PRs anyways.

**Comment by simon-mo** [comment]
On the worker killed part, we actually _can't_ retry the query because user got the object ref waaay before the replica is killed. And Ray's retry mechanism won't work because workers are intentionally killed.

**Comment by edoakes** [comment]
@simon-mo good point, hadn't thought about the fact that we're returning the object ref directly. If we were to wrap the object ref in our own `ServeResult` object, we could transparently handle the retries.

That said, we should do graceful shutdown anyways, so I'm fine with that solution for now.

**Comment by simon-mo** [comment]
Ok new plan incorporating @edoakes's feedback:
- This PR will be closed. It will be replaced by:
1. Introduce long puller, and let it replace existing controller -> {router, replica} communication.
2. Refactor router into sync and async routers, wrap ServeHandle API on top. This will depend on 1. 
3. Fix the scale down issue by introduce graceful termination.

**Comment by edoakes** [comment]
@simon-mo that sounds reasonable, but does it mean we'll have a regression between 1 and 3 because we haven't addressed the issues with scale down? Any reason not to do that first?

**Comment by simon-mo** [comment]
Good point. I thought 3 actually depends on 2 but actually it did not. We can fix 3 independently.
